src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -1057,3 +1057,152 @@ cdef class DrawFuncs:
         self._close_path_func = func
         hb_draw_funcs_set_close_path_func(
             self._hb_drawfuncs, _close_path_func, <void*>user_data, NULL)
+
+cdef class HBLinks:
+    cdef hb_link_t* _hb_link
+
+    cdef set(self, hb_link_t* p_link):
+        self._hb_link = p_link
+
+    def add_links_value(self, links: List[Tuple[int, int, int]]):
+        cdef unsigned int num_links = len(links)
+        for i in range(num_links):
+            self._hb_link[i].position = links[i][0]
+            self._hb_link[i].width = links[i][1]
+            self._hb_link[i].objidx = links[i][2]
+
+cdef hb_link_t* create_links(unsigned int link_num,
+                             hb_object_t* obj_list,
+                             unsigned int length,
+                             hb_bool_t* success):
+    success[0] = 1
+    cdef hb_link_t* p = NULL
+    if link_num == 0:
+        return NULL
+    else:
+        p = <hb_link_t*>calloc(link_num, sizeof(hb_link_t))
+        if p == NULL:
+            free_allocated(obj_list, length)
+            # for virtual_links, need to free additional real_links which
+            # is created before virtual_links
+            if obj_list[length].real_links != NULL:
+                free(obj_list[length].real_links)
+            free(obj_list)
+
+            success[0] = 0
+            return NULL
+        else:
+            return p
+
+cdef void free_allocated(hb_object_t* obj_list, unsigned int length):
```

Comment:
cython `cdef class` classes can have a `__dealloc__` method that is called when the object is garbage collected, and is used to free heap allocated memory, e.g. see

https://github.com/fonttools/skia-pathops/blob/0728fbdbf61b73a32207849026850a493b534cac/src/python/pathops/_pathops.pyx#L983-L996

the idea is you define a custom python type (cdef class) that can _own_ pointers to raw memory whose lifetime is tied to that of the python object itself, so you don't have to manually call malloc and free.
You can allocate the memory in the special `__cinit__` method (the default constructor for cdef class types which is called with the same parameters as the python `__init__` and where c stuff can be initialized) or define alternative class constructors (decorated with @staticmethod like the `_VerbArray.create` I linked to above) that return an instance of this cdef class, use it and then let the garbage collector call `__dealloc__` for you automatically when refcount is 0.